{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# coding: utf-8\n",
    "\n",
    "\"\"\"\n",
    "2020/08/26\n",
    "author:guisen chen\n",
    "email:thecgs001@foxmail.com\n",
    "description: By genblastG software, homologous gene of target \n",
    "             species was predicted based on homologous protein \n",
    "             sequences, and GFF files was generated.\n",
    "usage:./run_genblastG.py seq.fa genome.fa\n",
    "\"\"\"f\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import subprocess\n",
    "from Bio import SeqIO\n",
    "\n",
    "os.chdir('/home/guisen/temp/genblast') # 设置工作路径\n",
    "print(f'当前工作目录为：{os.getcwd()}')\n",
    "\n",
    "def get_single_seq(seq_file):\n",
    "    \"\"\"\n",
    "    将multiple_seq割成以seq_id.fa为文件名的single_seq文件,并返回一个以seq_id为元素的list\n",
    "    \"\"\"\n",
    "    seq_records = open(seq_file,'r')\n",
    "    id_list = []\n",
    "    for seq_record in SeqIO.parse(seq_records,'fasta'):\n",
    "        id_list.append(seq_record.id)\n",
    "        single_seq_file = open(f'{seq_record.id}.fa','w')\n",
    "        single_seq_file.write(f'>{seq_record.id}\\n{seq_record.seq}')\n",
    "        single_seq_file.close()\n",
    "    seq_records.close()\n",
    "    return id_list\n",
    "    \n",
    "def run_gblast(single_seq,genome_file):\n",
    "    \"\"\"\n",
    "    运行gblastG,并生成gene_cds，gene_pep，gene_gff文件，此软件的当前目录必须有alignscore.txt与blastall两个文件\n",
    "    \"\"\"\n",
    "    cline = subprocess.Popen(f'genblastG -P blast -q {single_seq} -t {genome_file} -e 1e-10 -r 5 -c 0.8 -gff -o gblast.out',env = None,shell=True)\n",
    "    cline.wait()\n",
    "    gene_gff = open('gene_gff','a')\n",
    "    headle_gff = open('gblast.out_1.1c_2.3_s1_0_16_1.gff','r')\n",
    "    temp_gff = headle_gff.read()\n",
    "    gene_gff.write(temp_gff)\n",
    "    gene_gff.close()\n",
    "    headle_gff.close()\n",
    "    os.remove('gblast.out_1.1c_2.3_s1_0_16_1')   # 此文件多序列文件运行会很大，所以每次运行一个序列后都删除\n",
    "    os.remove('gblast.out_1.1c_2.3_s1_0_16_1.gff')\n",
    "    os.remove(f'{single_seq}_{os.path.basename(genome_file)}.blast')\n",
    "    os.remove(f'{single_seq}_{os.path.basename(genome_file)}.blast.report')\n",
    "    os.remove('perform.txt')\n",
    "    \n",
    "def reconsitution_gff(file_gff):\n",
    "    \"\"\"\n",
    "    清洗file.gff\n",
    "    \"\"\"\n",
    "    gene_gff = open(f'{file_gff}','r')\n",
    "    \n",
    "    #删除注释行，将gff_file的第3列的transcript改为mRNA，coding_exon改为CDS，并返回一个以行为元素的clean_list\n",
    "    clean_line = []\n",
    "    for line in gene_gff:\n",
    "        if line[0] != '#':\n",
    "            clean_line.append(line.split())\n",
    "            \n",
    "            for line_list in clean_line:\n",
    "                if line_list[2].strip() == 'transcript':\n",
    "                    line_list[2] = 'gene'\n",
    "                elif line_list[2].strip() == 'coding_exon':\n",
    "                    line_list[2] = 'exon'                \n",
    "    gene_gff.close()\n",
    "    \n",
    "    #返回一个以gene_id为元素的id_list,用于cluster\n",
    "    id_list = []\n",
    "    for line_list in clean_line:\n",
    "        if line_list[2].strip() == 'gene':\n",
    "            id_list.append(line_list[8].split(';')[0][3:])\n",
    "    id_list.reverse()\n",
    "    \n",
    "    #返回以一个三重嵌套list,第一层为gene_structure_cluster\n",
    "    cluster_list = []\n",
    "    while len(id_list) > 0:\n",
    "        id_name = id_list.pop()\n",
    "        temp_list=[]\n",
    "        for line_list in clean_line:\n",
    "            if line_list[8].find(id_name) > 0:\n",
    "                temp_list.append(line_list)\n",
    "        cluster_list.append(temp_list)\n",
    "\n",
    "    #改为标准的gff3格式，并对cluster以染色体顺序排序,返回cluster_list\n",
    "    for cluster in cluster_list:\n",
    "        for line_list in cluster:\n",
    "            #print(line_list)\n",
    "            if line_list[2] == 'gene':\n",
    "                id = str(f'{line_list[8].split(\";\")[0][3:]}')\n",
    "                name = str(line_list[8].split(';')[1][5:])\n",
    "                line_list[8] = f'ID={id};Name={name}'\n",
    "            elif line_list[2] == 'exon':\n",
    "                line_list[8] = f'ID={id}.exon{cluster.index(line_list)};Parent={id}'\n",
    "    cluster_list.sort()\n",
    "    \n",
    "    for cluster in cluster_list:\n",
    "        for line_list in cluster:\n",
    "            if line_list[2] == 'gene':\n",
    "                id = str(f'{line_list[8].split(\";\")[0][3:]}')\n",
    "                add_list = line_list.copy()\n",
    "                add_list[2] = 'mRNA'\n",
    "                add_list[8] = f'ID=mrna.{id};Parent={id}'\n",
    "                add_list[5] = '.'\n",
    "                cluster.insert(cluster.index(line_list)+1,add_list)\n",
    "            elif line_list[2] == 'exon':\n",
    "                add_list = line_list.copy()\n",
    "                add_list[2] = 'CDS'\n",
    "                add_list[8] = f'ID=cds.{id};Parent={id}'\n",
    "                cluster.insert(cluster.index(line_list)+1,add_list)\n",
    "    return cluster_list\n",
    "    \n",
    "#执行\n",
    "id_list = get_single_seq(sys.argv[1]) \n",
    "for id in id_list:\n",
    "    run_gblast(f'{id}.fa',sys.argv[2])\n",
    "    os.remove(f'{id}.fa') # 删除生成的single_seq_file\n",
    "    \n",
    "#写入gff文件\n",
    "gff_file = open('./gene.gff','a')\n",
    "gff = csv.writer(gff_file,delimiter='\\t')\n",
    "for cluster in reconsitution_gff('./gene_gff'):\n",
    "    for line_list in cluster:\n",
    "        gff.writerow(line_list)\n",
    "gff_file.close()\n",
    "os.remove('./gene_gff')\n",
    "print('-------completed-------')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
