{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[38689495, 38690967, 38689173]\t[38689620, 38691080, 38689337]\n",
      "evm.model.Lachesis13__1343704387.1154\n",
      "38689173\n",
      "38691080\n",
      "提取第1个Hit成功\n",
      "[35120180, 35122182]\t[35120509, 35122385]\n",
      "evm.model.Lachesis12__1644025036.1116\n",
      "35120180\n",
      "35122385\n",
      "提取第2个Hit成功\n",
      "[40339545]\t[40339772]\n",
      "evm.model.Lachesis12__1644025036.1116\n",
      "40339545\n",
      "40339772\n",
      "提取第3个Hit成功\n",
      "[9039068]\t[9039298]\n",
      "evm.model.Lachesis2__1753060648.194\n",
      "9039068\n",
      "9039298\n",
      "提取第4个Hit成功\n",
      "[9064962, 9064636]\t[9065204, 9064794]\n",
      "evm.model.Lachesis2__1753060648.195\n",
      "9064636\n",
      "9065204\n",
      "提取第5个Hit成功\n",
      "[49360120, 49363020, 49363376]\t[49360356, 49363283, 49363438]\n",
      "evm.model.Lachesis2__1753060648.1294\n",
      "49360120\n",
      "49363438\n",
      "提取第6个Hit成功\n",
      "[35120285]\t[35120509]\n",
      "evm.model.Lachesis2__1753060648.1294\n",
      "35120285\n",
      "35120509\n",
      "提取第7个Hit成功\n",
      "[40337043, 40339548]\t[40337321, 40339796]\n",
      "evm.model.Lachesis5__547510254.1454\n",
      "40337043\n",
      "40339796\n",
      "提取第8个Hit成功\n",
      "[35120282]\t[35120485]\n",
      "evm.model.Lachesis5__547510254.1454\n",
      "35120282\n",
      "35120485\n",
      "提取第9个Hit成功\n",
      "completed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\n'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "\n",
    "\"\"\"\n",
    "批量在NCBI与SWISSPRO下载同源蛋白\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "同源蛋白去冗余\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "tblastn生成XML文件\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "提取XML文件hsp.sbjct氨基酸序列，生成sbjct_protein文件\n",
    "根据XML文件的hit_id提取基因组hit_macth区上下游1500bp的DNA，生成sbjct_genome文件\n",
    "\"\"\"\n",
    "handle = open('/home/guisen/IL/IL-17-cal.xml','rt') #加载XML文件\n",
    "blast_results = NCBIXML.parse(handle)#解析\n",
    "i = 0#自增变量初始值\n",
    "list_num = []#自增变量列表收集\n",
    "for blast_record in blast_results:\n",
    "    for alignment in blast_record.alignments:\n",
    "        hsp_stat = []\n",
    "        hsp_end = []\n",
    "        for hsp in alignment.hsps:\n",
    "            hsp_stat.append(hsp.sbjct_start)\n",
    "            hsp_end.append(hsp.sbjct_end)\n",
    "        print(f'{hsp_stat}\\t{hsp_end}')\n",
    "        print(blast_record.query)\n",
    "        for seq_record in SeqIO.parse('/home/guisen/03.structure_annotation/genome.fa','fasta'):\n",
    "            if alignment.hit_id.split('|')[0].strip() == seq_record.id.strip():\n",
    "                print(min(hsp_stat))\n",
    "                print(max(hsp_end))\n",
    "                if min(hsp_stat) - 1500 < 0:\n",
    "                    seq_start = 0\n",
    "                else:\n",
    "                    seq_start = min(hsp_stat) - 1500\n",
    "                if max(hsp_end) + 1500 < len(seq_record.seq):\n",
    "                        seq_end = max(hsp_stat) + 1500\n",
    "                else:\n",
    "                    seq_end = len(seq_record.seq)\n",
    "                i += 1\n",
    "                list_num.append(i)\n",
    "                #sbjct_protein = open(f'/home/guisen/tmp/sbjct_protein{i}.fa','w+')\n",
    "                #sbjct_protein.write(f'>query_id:{blast_record.query}|alignment_id:{alignment.title}|strat:{hsp.sbjct_start}|end:{hsp.sbjct_end}|e_value:{hsp.expect}|score:{hsp.score}\\n{hsp.sbjct}')\n",
    "                #sbjct_protein.close()\n",
    "                sbjct_genome = open(f'/home/guisen/tmp/sbjct_genome{i}.fa','w+')\n",
    "                sbjct_genome.write(f'>{seq_record.id}\\n{seq_record.seq[seq_start:seq_end]}')\n",
    "                sbjct_genome.close()\n",
    "                print(f'提取第{i}个Hit成功')#显示\n",
    "            else:\n",
    "                pass\n",
    "        for seq_protein in SeqIO.parse('/home/guisen/IL/IL-17-cal.fa','fasta'):\n",
    "                if blast_record.query == seq_protein.id:\n",
    "                    sbjct_protein = open(f'/home/guisen/tmp/sbjct_protein{i}.fa','w+')\n",
    "                    sbjct_protein.write(f'>query_id:{blast_record.query}\\n{seq_protein.seq}')\n",
    "                    sbjct_protein.close()\n",
    "                                       \n",
    "\"\"\"\n",
    "并行化genewise，批量预测基因\n",
    "\"\"\"\n",
    "for i in list_num:\n",
    "    cline = subprocess.Popen(f'genewise sbjct_protein{i}.fa sbjct_genome{i}.fa -both -gff -cdna -pep -genes>> sbjct_predicted.gff',env={'PATH':'/opt/biosoftware/wise2.4.1/src/bin','WISECONFIGDIR':'/opt/biosoftware/wise2.4.1/wisecfg/'},shell=True,cwd='/home/guisen/tmp/')\n",
    "print('completed')\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
